### PR TITLE
Android: fix GUI not loading on Qt 5.15

### PR DESCRIPTION
**Pull request by icota**
Qt 5.14 introduced certain [breaking changes](https://lists.qt-project.org/pipermail/interest/2020-January/034372.html) to the way it parses `AndroidManifest.xml` metadata.

This PR also adds an explicit `QAndroidPlatformIntegrationPlugin` import to `bitcoin.cpp` (in line with other platforms). ~~I'm not sure why Android GUI worked without this before but it certainly doesn't on Qt 5.15.~~

We could do without `Q_IMPORT_PLUGIN` before because we were manually pulling a `JNI_Onload` function to the linkage.

### CODE REVIEW

**Comment by icota** [comment]
cc @hebasto

**Comment by hebasto** [comment]
Concept ACK.

**Comment by hebasto** [comment]
https://cirrus-ci.com/task/5550307582148608:
```
ld: error: duplicate symbol: JNI_OnLoad
>>> defined at qjnionload.cpp
>>>            qjnionload.o:(JNI_OnLoad) in archive /tmp/cirrus-ci-build/depends/aarch64-linux-android/lib/libQt5Core_arm64-v8a.a
>>> defined at androidjnimain.cpp
>>>            androidjnimain.o:(.text.JNI_OnLoad+0x0) in archive /tmp/cirrus-ci-build/depends/aarch64-linux-android/plugins/platforms/libplugins_platforms_qtforandroid_arm64-v8a.a
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:6462: qt/test/test_bitcoin-qt] Error 1
```

**Comment by hebasto** [comment]
Tested f2f924e919b110cc5f33aac4caa4b296991d3706 (configured with `--disable-tests`):

Screenshot_1639304796

**Comment by icota** [comment]
I've updated the PR description to explain why we could get away with not using `Q_IMPORT_PLUGIN`. 

Broken `qt_test` linkage in CI is related. I'm trying to figure out a nice way to fix this.

**Comment by hebasto** [comment]
> I've updated the PR description to explain why we could get away with not using `Q_IMPORT_PLUGIN`.
> 
> Broken `qt_test` linkage in CI is related. I'm trying to figure out a nice way to fix this.

The following patch fixes link error:
```diff
--- a/build-aux/m4/bitcoin_qt.m4
+++ b/build-aux/m4/bitcoin_qt.m4
@@ -162,6 +162,7 @@ AC_DEFUN([BITCOIN_QT_CONFIGURE],[
       AC_DEFINE([QT_QPA_PLATFORM_COCOA], [1], [Define this symbol if the qt platform is cocoa])
     elif test "$TARGET_OS" = "android"; then
       QT_LIBS="-Wl,--export-dynamic,--undefined=JNI_OnLoad -lplugins_platforms_qtforandroid_$ANDROID_ARCH -ljnigraphics -landroid -lqtfreetype_$ANDROID_ARCH $QT_LIBS"
+      QT_TEST_LIBS="-Wl,--export-dynamic,--undefined=JNI_OnLoad -lplugins_platforms_qtforandroid_$ANDROID_ARCH -ljnigraphics -landroid -lqtfreetype_$ANDROID_ARCH $QT_TEST_LIBS"
       AC_DEFINE([QT_QPA_PLATFORM_ANDROID], [1], [Define this symbol if the qt platform is android])
     fi
   fi
```

Not sure if it is optimal though.

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -65,6 +65,8 @@ Q_IMPORT_PLUGIN(QWindowsVistaStylePlugin);
 #elif defined(QT_QPA_PLATFORM_COCOA)
 Q_IMPORT_PLUGIN(QCocoaIntegrationPlugin);
 Q_IMPORT_PLUGIN(QMacStylePlugin);
+#elif defined(QT_QPA_PLATFORM_ANDROID)
+Q_IMPORT_PLUGIN(QAndroidPlatformIntegrationPlugin);
```
Comment (hebasto) [inline src/qt/bitcoin.cpp:69]: The same in `qt/test/test_main.cpp`?

**Comment by hebasto** [comment]
> > I've updated the PR description to explain why we could get away with not using `Q_IMPORT_PLUGIN`.
> > Broken `qt_test` linkage in CI is related. I'm trying to figure out a nice way to fix this.
> 
> The following patch fixes link error:
> 
> ```diff
> --- a/build-aux/m4/bitcoin_qt.m4
> +++ b/build-aux/m4/bitcoin_qt.m4
> @@ -162,6 +162,7 @@ AC_DEFUN([BITCOIN_QT_CONFIGURE],[
>        AC_DEFINE([QT_QPA_PLATFORM_COCOA], [1], [Define this symbol if the qt platform is cocoa])
>      elif test "$TARGET_OS" = "android"; then
>        QT_LIBS="-Wl,--export-dynamic,--undefined=JNI_OnLoad -lplugins_platforms_qtforandroid_$ANDROID_ARCH -ljnigraphics -landroid -lqtfreetype_$ANDROID_ARCH $QT_LIBS"
> +      QT_TEST_LIBS="-Wl,--export-dynamic,--undefined=JNI_OnLoad -lplugins_platforms_qtforandroid_$ANDROID_ARCH -ljnigraphics -landroid -lqtfreetype_$ANDROID_ARCH $QT_TEST_LIBS"
>        AC_DEFINE([QT_QPA_PLATFORM_ANDROID], [1], [Define this symbol if the qt platform is android])
>      fi
>    fi
> ```
> 
> Not sure if it is optimal though.

A more optimal suggestion:
```diff
--- a/src/Makefile.qttest.include
+++ b/src/Makefile.qttest.include
@@ -52,7 +52,7 @@ if ENABLE_ZMQ
 qt_test_test_bitcoin_qt_LDADD += $(LIBBITCOIN_ZMQ) $(ZMQ_LIBS)
 endif
 qt_test_test_bitcoin_qt_LDADD += $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) $(LIBLEVELDB) \
-  $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BOOST_LIBS) $(QT_DBUS_LIBS) $(QT_TEST_LIBS) $(QT_LIBS) \
+  $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BOOST_LIBS) $(QT_LIBS) $(QT_DBUS_LIBS) $(QT_TEST_LIBS) \
   $(QR_LIBS) $(BDB_LIBS) $(MINIUPNPC_LIBS) $(NATPMP_LIBS) $(LIBSECP256K1) \
   $(EVENT_PTHREADS_LIBS) $(EVENT_LIBS) $(SQLITE_LIBS)
 qt_test_test_bitcoin_qt_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(QT_LDFLAGS) $(LIBTOOL_APP_LDFLAGS) $(PTHREAD_FLAGS)
```

File: src/qt/android/.gitignore [hunk 2]
```diff
@@ -3,7 +3,5 @@
 /gradle/wrapper
 /gradlew*
 /libs
-/res/layout
-/res/values*
```
Comment (hebasto) [inline src/qt/android/.gitignore:7]: While here, maybe s|`/gradle/wrapper`|`/gradle`| as `/gradle/wrapper` is the only subdirectory in `/gradle`, no?

**Comment by icota** [comment]
> UPDATE: Suggesting to move this PR into the main repo as it touches the code outside the `src/qt`.

Is there a nice way to do this? I'd like to preserve comments and stuff.

**Comment by hebasto** [comment]
@icota 
> > UPDATE: Suggesting to move this PR into the main repo as it touches the code outside the `src/qt`.
> 
> Is there a nice way to do this?

No, there is not, unfortunately.

> I'd like to preserve comments and stuff.

A link to this PR in the OP will work fine.

**Comment by icota** [comment]
Continued in https://github.com/bitcoin/bitcoin/pull/23757

**Pull request closed without merging**